catchup/service.go

```diff
@@ -463,105 +430,86 @@ func (s *Service) pipelinedFetch(seedLookback uint64) {
 		parallelRequests = seedLookback
 	}
 
-	completed := make(chan basics.Round, parallelRequests)
-	taskCh := make(chan task, parallelRequests)
+	completed := make(map[basics.Round]chan bool)
 	var wg sync.WaitGroup
-
 	defer func() {
-		close(taskCh)
 		wg.Wait()
-		close(completed)
+		for _, ch := range completed {
+			close(ch)
+		}
 	}()
 
 	peerSelector := createPeerSelector(s.net, s.cfg, true)
-
 	if _, err := peerSelector.getNextPeer(); err == errPeerSelectorNoPeerPoolsAvailable {
 		s.log.Debugf("pipelinedFetch: was unable to obtain a peer to retrieve the block from")
 		return
 	}
 
-	// Invariant: len(taskCh) + (# pending writes to completed) <= N
-	wg.Add(int(parallelRequests))
-	for i := uint64(0); i < parallelRequests; i++ {
-		go func() {
-			defer wg.Done()
-			for t := range taskCh {
-				completed <- t() // This write comes after a read from taskCh, so the invariant is preserved.
+	// Create a new context for canceling the pipeline if some block
+	// fetch fails along the way.
+	ctx, cancelCtx := context.WithCancel(s.ctx)
+	defer cancelCtx()
+
+	// firstRound is the first round we're waiting to fetch.
+	firstRound := s.ledger.NextRound()
+
+	// nextRound is the next round that we will issue a fetch for.
+	nextRound := firstRound
+
+	for {
+		for nextRound < firstRound+basics.Round(parallelRequests) {
+			if s.nextRoundIsNotSupported(nextRound) {
+				// We may get here when (1) The service starts
+				// and gets to an unsupported round.  Since in
+				// this loop we do not wait for the requests
+				// to be written to the ledger, there is no
+				// guarantee that the unsupported round will be
+				// stopped in this case.
```

Comment:
This comment is good to keep.

```suggestion
				// We may get here when the service starts
				// and gets to an unsupported round.  Since in
				// this loop we do not wait for the requests
				// to be written to the ledger, there is no
				// guarantee that the unsupported round will be
				// stopped in this case.
```

Reply:
Thank you!